### PR TITLE
Add a runner-agnostic client for WAB servers.

### PR DESCRIPTION
**Pull request by ohler55**
Implements a client that interacts with a WAB server. This is useful for interacting with the WAB server from either an application of from `irb` during debugging.

### CODE REVIEW

File: test/client_test/wabur.conf [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+# the site directory.
+export_proxy = true
+
+# The number of spaces to indent the JSON written to disk by the model and the
+# JSON sent as a response to View requests.
+indent = 0
+
+# Handlers for each type of record and for the UI shich are Ruby generated UI
+# configuration records.
+handler.0.type =
```
Comment (ashmaroli) [inline test/client_test/wabur.conf]: empty declaration?

File: lib/wab/client.rb [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+          content = content.json
+        else
+          content = Oj.dump(content, mode: :wab, indent: 0)
+        end
+        header['Content-Type'] = 'application/json'
+        resp = @http.send_request(method, form_path(kind, query), content, header)
+      else
+        resp = @http.send_request(method, form_path(kind, query))
+      end
+      raise Error.new(resp.body) if Net::HTTPOK != resp.class
```
Comment (ashmaroli) [inline lib/wab/client.rb]:
```ruby
raise Error.new(resp.body) unless Net::HTTPOK == resp.class
```
or
```ruby
raise Error.new(resp.body) unless Net::HTTPOK === resp
```

File: lib/wab/client.rb [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        path << '/'
+        path << query.to_s
+      end
+      path
+    end
+    
+    def send_request(method, kind, query, content)
+      connect
+      header = {'Connection' => (@keep_alive ? 'Keep-Alive' : 'Close') }
+      unless content.nil?
```
Comment (ashmaroli) [inline lib/wab/client.rb]: Rubocop would've prompted you to not use `unless..else` statements.. :smiley: 
Instead it'll tell you to re-arrange the branches to `if..else`
```ruby
if content.nil?
  resp = @http.send_request(method, form_path(kind, query))
else
  # ...
end
```

File: lib/wab/client.rb [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+      raise ArgError.new('tql') if tql.nil?
+      send_request('POST', 'tql', nil, tql)
+    end
+    
+    private
+
+    def connect
+      return unless @http.nil?
+      @http = Net::HTTP.new(@server_address, @server_port)
+    end
```
Comment (ashmaroli) [inline lib/wab/client.rb]: Would this be the same as:
```ruby
def connect
  @http ||= Net::HTTP.new(@server_address, @server_port)
end

File: lib/wab/client.rb [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+    # Create a new Client for the server at +addr+ and +port+. The provided
+    # options should match the attribute names and types.
+    def initialize(addr, port, options={})
+      @server_address = addr
+      @server_port = port
+      @path_prefix = options.fetch(:path_prefix, '/v1/')
+      @tql_path = options.fetch(:tql_path, '/tql')
+      @type_key = options.fetch(:type_key, 'kind')
+      @keep_alive = options.fetch(:keep_alive, true)
```
Comment (ashmaroli) [inline lib/wab/client.rb]: Some developers do not like using double negation `!!` but wouldn't this be better if it was..: ?
```ruby
@keep_alive = !!options.fetch(:keep_alive, true)
```

File: lib/wab/client.rb (hunk 3 above)
Comment (ohler55) [reply lib/wab/client.rb]: I don't think it really matters.

File: lib/wab/client.rb (hunk 2 above)
Comment (ohler55) [reply lib/wab/client.rb]: Has to be the class. Could use `is_a?` but has to be the class.

File: test/client_test/wabur.conf (hunk 1 above)
Comment (ohler55) [reply test/client_test/wabur.conf]: yes, that is the wild card, catch everything notation.

File: lib/wab/client.rb (hunk 2 above)
Comment (ashmaroli) [reply lib/wab/client.rb]: > Has to be the class.

I know using `===` is the same as `is_a?`. I prefer `is_a?` too..

Comment (ohler55) [reply lib/wab/client.rb]: I don't care much either way. Changed.

**Comment by ohler55** [comment]
It looks like the older waburs are not being killed. Maybe different port would work around that.

**Pull request merged**